modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1127 @@
+{
+  // TODO: Need this for branching in 3D
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::intersectWithEdge(const Point & p1,
+                                       const Point & p2,
+                                       const std::vector<Point> & vertices,
+                                       Point & pint) const
+{
+  bool has_intersection = false;
+
+  Point normal;
+  Point center;
+  findElemNormal(vertices, normal, center);
+
+  std::vector<Real> plane_point = {center(0), center(1), center(2)};
+  std::vector<Real> planenormal = {normal(0), normal(1), normal(2)};
+  std::vector<Real> edge_point1 = {p1(0), p1(1), p1(2)};
+  std::vector<Real> edge_point2 = {p2(0), p2(1), p2(2)};
+  std::vector<Real> cut_point = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(
+          &plane_point[0], &planenormal[0], &edge_point1[0], &edge_point2[0], &cut_point[0]) == 1)
+  {
+    Point temp_p(cut_point[0], cut_point[1], cut_point[2]);
+    if (isInsideCutPlane(vertices, temp_p) && isInsideEdge(p1, p2, temp_p))
+    {
+      pint = temp_p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::findIntersection(const Point & p1,
+                                      const Point & p2,
+                                      const std::vector<Point> & vertices,
+                                      Point & pint) const
+{
+  bool has_intersection = false;
+
+  Point normal;
+  Point center;
+  findElemNormal(vertices, normal, center);
+
+  std::vector<Real> plane_point = {center(0), center(1), center(2)};
+  std::vector<Real> planenormal = {normal(0), normal(1), normal(2)};
+  std::vector<Real> p_begin = {p1(0), p1(1), p1(2)};
+  std::vector<Real> p_end = {p2(0), p2(1), p2(2)};
+  std::vector<Real> cut_point = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(
+          &plane_point[0], &planenormal[0], &p_begin[0], &p_end[0], &cut_point[0]) == 1)
+  {
+    Point p(cut_point[0], cut_point[1], cut_point[2]);
+    Real dotp = ((p - p1) * (p2 - p1)) / ((p2 - p1) * (p2 - p1));
+    if (isInsideCutPlane(vertices, p) && dotp > 1)
+    {
+      pint = p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::isInsideEdge(const Point & p1, const Point & p2, const Point & p) const
+{
+  Real dotp1 = (p1 - p) * (p2 - p1);
+  Real dotp2 = (p2 - p) * (p2 - p1);
+  return (dotp1 * dotp2 <= 0.0);
+}
+
+Real
+MeshCut3DUserObject::getRelativePosition(const Point & p1, const Point & p2, const Point & p) const
+{
+  Real full_len = (p2 - p1).norm();
+  Real len_p1_p = (p - p1).norm();
+  return len_p1_p / full_len;
+}
+
+bool
+MeshCut3DUserObject::isInsideCutPlane(const std::vector<Point> & vertices, Point p) const
+{
+  unsigned int n_node = vertices.size();
+
+  Point normal;
+  Point center;
+
+  findElemNormal(vertices, normal, center);
```

Comment:
We can already fine normals to planes. It's in libMesh.

Reply:
Made the change and removed this method.